### PR TITLE
feat: added NewAssignmentNotification - to all groupmembers

### LINKED ISSUE #3294: feat: notify group members for the new assignment in the group
**Is your feature request related to a problem? Please describe.**
- All group members of a group should get notify regarding the new assignment in the group.

**Describe the solution you'd like**
- Create a new notification in `app/notifications`(we are using the `noticed` gem, [documentation](https://github.com/excid3/noticed))
- When group mentors click the submit button in the form for a new assignment, the notifications should be delivered to all group members as recipients.

**Are you working on this? (Yes/No)**
No

### PR DESCRIPTION
**Pull request by VaibhavUpreti**
Fixes #3294 

#### Describe the changes you have made in this PR -
- added notifications to group members on new Assignment creation
### Screenshots of the changes (If any) -

Screenshot 2022-10-10 at 7 44 42 PM

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by VaibhavUpreti** [comment]
@vedant-jain03 kindly review.

File: app/models/assignment.rb [hunk 1]
```diff
@@ -8,15 +8,33 @@ class Assignment < ApplicationRecord
...
 
+  has_noticed_notifications model_name: "NoticedNotification"
+  has_many :noticed_notifications, through: :author
+  has_many :notifications, as: :notifiable  # rubocop:disable Rails/HasManyOrHasOneDependent
+
+  def notify_recipient
+    @assignment = Assignment.find(id)
+    group.group_members.each do |group_member|
+      # NewAssignmentNotification.with(user: group_member.user, assingment: @assingment).deliver(group_member.user)
+      NewAssignmentNotification.with(assingment: @assignment).deliver(group_member.user)
```
Comment (vedant-jain03) [inline app/models/assignment.rb:30]:
```suggestion
      NewAssignmentNotification.with(assignment: @assignment).deliver_later(group_member.user)
```

File: app/models/assignment.rb [hunk 2]
```diff
@@ -8,15 +8,33 @@ class Assignment < ApplicationRecord
...
   after_commit :send_update_mail, on: :update
+  after_create_commit :notify_recipient
 
   enum grading_scale: { no_scale: 0, letter: 1, percent: 2, custom: 3 }
   default_scope { order(deadline: :asc) }
   has_many :grades, dependent: :destroy
 
+  has_noticed_notifications model_name: "NoticedNotification"
+  has_many :noticed_notifications, through: :author
+  has_many :notifications, as: :notifiable  # rubocop:disable Rails/HasManyOrHasOneDependent
```
Comment (vedant-jain03) [inline app/models/assignment.rb:24]: These associations are not needed, 
`has_noticed_notifications model_name: "NoticedNotification"` will do these tasks!
Remove this!

File: db/schema.rb [hunk 3]
```diff
@@ -20,7 +19,7 @@
     t.bigint "user_id"
     t.string "name"
     t.json "properties"
-    t.datetime "time"
+    t.datetime "time", precision: nil
```
Comment (vedant-jain03) [inline db/schema.rb:22]: Remove these changes

File: app/controllers/users/noticed_notifications_controller.rb [hunk 4]
```diff
@@ -12,6 +12,7 @@ def mark_as_read
     notification = NoticedNotification.find(params[:notification_id])
     notification.update(read_at: Time.zone.now)
     project = notification.params[:project]
+    # assignment = notification.params[:assignment]
     redirect_to user_project_path(project.author, project)
```
Comment (vedant-jain03) [inline app/controllers/users/noticed_notifications_controller.rb:16]:
```suggestion
    if notification.type == "NewAssignmentNotification"
      assignment = notification.params[:assignment]
      group = assignment.group
      redirect_to group_assignment_path(group, assignment)
    else
      project = notification.params[:project]
      redirect_to user_project_path(project.author, project)
    end
```

This is for redirection purpose, when user clicks to notification where it should get redirected, previously we had star and fork notification for which user should get redirected to project path, but for assignment it should get redirect to assignment path which can be achieved using this.

Test it!

File: config/locales/views/users/en.yml [hunk 5]
```diff
@@ -91,3 +91,4 @@ en:
       see_all : "See All"
       fork_notification: "%{user} forked your Project %{project}"
       star_notification: "%{user} starred your Project %{project}"
+      new_assignment_notification: "You have been assigned a new assignment"
```
Comment (vedant-jain03) [inline config/locales/views/users/en.yml:94]:
```suggestion
      new_assignment_notification: "You have been assigned a new assignment '%{assignment_name}'"
```

File: app/notifications/new_assignment_notification.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class NewAssignmentNotification < Noticed::Base
+  deliver_by :database, association: :noticed_notifications
+
+  def message
+    # user = params[:group_member.user]
+    assignment = params[:assignment]
+    #  t("users.notifications.new_assignment_notification", assignment: assingment.name)
+    t("users.notifications.new_assignment_notification", assignment: assignment)
```
Comment (vedant-jain03) [inline app/notifications/new_assignment_notification.rb:10]:
```suggestion
    t("users.notifications.new_assignment_notification", assignment_name: assignment.name)
```

File: config/locales/views/users/hi.yml [hunk 7]
```diff
@@ -90,4 +90,5 @@ hi:
       no_unread_notifications: "आपके पास कोई अपठित नोटिफिकेशन्स नहीं है!"
       see_all : "सभी देखें"
       fork_notification: "%{user} ने आपके प्रोजेक्ट %{project} को फोर्क किया"
-      star_notification: "%{user} ने आपका प्रोजेक्ट %{project} स्टार किया"
\ No newline at end of file
+      star_notification: "%{user} ने आपका प्रोजेक्ट %{project} स्टार किया"
+      new_assignment_notification: "आपको एक नया कार्य सौंपा गया है!"
```
Comment (vedant-jain03) [inline config/locales/views/users/hi.yml:94]:
```suggestion
      new_assignment_notification: "आपको एक नया कार्य '%{assignment_name}' सौंपा गया है!"
```

File: app/notifications/new_assignment_notification.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  def icon
+    "far fa-star fa-thin"
```
Comment (vedant-jain03) [inline app/notifications/new_assignment_notification.rb:14]:
```suggestion
    "fa fa-clipboard"
```

Try this one
cc @nitin10s

**Review by vedant-jain03 — CHANGES_REQUESTED** [review]
Thanks @VaibhavUpreti , here are my reviews!
Kindly go through and let me know if you have any doubt!

cc @tachyons

**Comment by VaibhavUpreti** [comment]
Hey @vedant-jain03 , thank you for the review.
I found some things were broken .
1. All pages were giving name error... due to new_assignment_notification.rb file .... 
name-error
Error seems to be in calling the method
    ''' assignment = params[:assignment] ''' 
I changed this to as shown in the image  
Untitled

After this ... all the name errors were gone ... and notifications were showing
this produces the following output
id

2. For the routing part
Screenshot 2022-10-15 at 2 11 36 AM
It was giving the following error
I think a possible fix for this could be to find group by params id. But alas it was still not working .. rather giving 404

File: app/notifications/new_assignment_notification.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class NewAssignmentNotification < Noticed::Base
+  deliver_by :database, association: :noticed_notifications
+
+  def message
+    # user = params[:group_member.user]
+    #assignment = params[:assignment]
+    assignment = Assignment.all
```
Comment (vedant-jain03) [inline app/notifications/new_assignment_notification.rb:9]: why `Assignment.all`?
assignment is in params, right?
you will get it `assignment = params[:assignment]`

Comment (VaibhavUpreti) [reply app/notifications/new_assignment_notification.rb:9]: image
I get this error on every page when I use
 assignment = params[:assignment]

Comment (vedant-jain03) [reply app/notifications/new_assignment_notification.rb:9]: It seems you have deleted the assignment which was saved in `noticed_notifications`!
The simple thing you can do, (as you are working on development)
in terminal:
- `rails c`
- `NoticedNotification.destroy_all`

And try again this!

Also you can ping me on slack for these issues!

Comment (VaibhavUpreti) [reply app/notifications/new_assignment_notification.rb:9]: Thanking you for helping me out so much @vedant-jain03 . This fix worked... (although I had to delete all assignments too)
The redirects are working.
Screenshot 2022-10-16 at 1 04 14 PM
And notifications are also showing for new assignments
Screenshot 2022-10-16 at 1 19 15 PM

However ... the green dot was overlapping with notification as mentioned in #3313 ... I came up with a quick fix for now for this
Screenshot 2022-10-16 at 1 21 01 PM

File: app/controllers/users/noticed_notifications_controller.rb [hunk 10]
```diff
@@ -11,8 +11,17 @@ def index
...
     notification.update(read_at: Time.zone.now)
-    project = notification.params[:project]
-    redirect_to user_project_path(project.author, project)
+    if notification.type == "NewAssignmentNotification"
+      assignment = notification.params[:assignment]
+      #group = assignment.group
+      #group = notification.params[:group]
+      group = Group.find(params[:group_id])
+      redirect_to group_assignment_path(group, assignment)
+      #redirect_to assignment_show_path(group, assignment)
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/noticed_notifications_controller.rb:20]: @VaibhavUpreti could you please remove the comments

File: app/controllers/users/noticed_notifications_controller.rb [hunk 11]
```diff
@@ -11,8 +11,17 @@ def index
   def mark_as_read
     notification = NoticedNotification.find(params[:notification_id])
     notification.update(read_at: Time.zone.now)
-    project = notification.params[:project]
-    redirect_to user_project_path(project.author, project)
+    if notification.type == "NewAssignmentNotification"
+      assignment = notification.params[:assignment]
+      #group = assignment.group
+      #group = notification.params[:group]
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/noticed_notifications_controller.rb:17]: @VaibhavUpreti remove the comment

File: app/controllers/users/noticed_notifications_controller.rb [hunk 12]
```diff
@@ -11,8 +11,17 @@ def index
   def mark_as_read
     notification = NoticedNotification.find(params[:notification_id])
     notification.update(read_at: Time.zone.now)
-    project = notification.params[:project]
-    redirect_to user_project_path(project.author, project)
+    if notification.type == "NewAssignmentNotification"
+      assignment = notification.params[:assignment]
+      #group = assignment.group
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/noticed_notifications_controller.rb:16]: @VaibhavUpreti remove the comment

File: app/models/assignment.rb [hunk 13]
```diff
@@ -8,15 +8,31 @@ class Assignment < ApplicationRecord
...
   default_scope { order(deadline: :asc) }
   has_many :grades, dependent: :destroy
 
+  has_noticed_notifications model_name: "NoticedNotification"
+
+  def notify_recipient
+    @assignment = Assignment.find(id)
+    group.group_members.each do |group_member|
+      # NewAssignmentNotification.with(user: group_member.user, assingment: @assingment).deliver(group_member.user)
+      NewAssignmentNotification.with(assingment: @assignment).deliver(group_member.user)
```
Comment (vedant-jain03) [inline app/models/assignment.rb]:
```suggestion
      NewAssignmentNotification.with(assignment: @assignment).deliver_later(group_member.user)
```

- `assingment` spelling is not correct!
- use `deliver_later` instead

File: app/notifications/new_assignment_notification.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class NewAssignmentNotification < Noticed::Base
+  deliver_by :database, association: :noticed_notifications
+
+  def message
+    # user = params[:group_member.user]
+    assignment = params[:assignment]
+    # t("users.notifications.new_assignment_notification", assignment: assingment.name)
```
Comment (vedant-jain03) [inline app/notifications/new_assignment_notification.rb:9]: remove the comments

**Comment by vedant-jain03** [comment]
@VaibhavUpreti could you please add `spec` testcases for this notifications!

Also test if the assignment got deleted, the associated notification are destroying or not!
That is the edge case, we need to see!

I think `has_noticed_notifications` do this thing but just in case kindly test this in your end!

**Comment by VaibhavUpreti** [comment]
> @VaibhavUpreti could you please add `spec` testcases for this notifications!
> 
> Also test if the assignment got deleted, the associated notification are destroying or not! That is the edge case, we need to see!
> 
> I think `has_noticed_notifications` do this thing but just in case kindly test this in your end!

Yeah Sure! I will write the required specs for these... And nothing seems to be broken after deleting the assignments ... The notifications are also getting cleaned up.

**Comment by VaibhavUpreti** [comment]
@vedant-jain03 I have written the tests ... can u please check .. also I had sent u dm on slack regarding testing controller

File: spec/models/assignment_spec.rb [hunk 15]
```diff
@@ -37,6 +37,14 @@
       end.to have_enqueued_job.on_queue("mailers")
     end
 
+    it "sends new assignment notification" do
+      expect do
+        @assignment = FactoryBot.create(:assignment, group: @group, status: "open")
+        NewAssignmentNotification.with(assignment: @assignment).deliver_later(group.first.user)
```
Comment (vedant-jain03) [inline spec/models/assignment_spec.rb:43]: Why this?

**Review by vedant-jain03 — CHANGES_REQUESTED** [review]

**Comment by vedant-jain03** [comment]
@VaibhavUpreti 
you need to write the test for this in `#create` in `assignment_controller_spec.rb`.

[Here](https://github.com/CircuitVerse/CircuitVerse/blob/d27a74baa875200e80733823e6be0d226d4e0ee6/spec/controllers/assignments_controller_spec.rb#L209)

Try adding under the context when primary mentor is logged in, try this:

```
      it "sends notifications to group members" do
        sign_in @primary_mentor
        post group_assignments_path(@group), params: { assignment:
          { description: "group assignment", name: "Test Name" } }
        expect(@member.noticed_notifications.count).to eq(1)
      end
```

File: spec/models/assignment_spec.rb (hunk 15 above)
Comment (VaibhavUpreti) [reply spec/models/assignment_spec.rb:43]: In this one I thought we could test the model function @assignment.notify_recipient .... whenever a new assignment is created .. sends notifications .

**Comment by VaibhavUpreti** [comment]
@vedant-jain03 I think this one if good to merge now.

**Review by vedant-jain03 — APPROVED** [review]
@VaibhavUpreti Great work, thanks!

cc @tachyons, we can merge this!

File: app/controllers/users/noticed_notifications_controller.rb [hunk 16]
```diff
@@ -11,8 +11,14 @@ def index
...
     notification.update(read_at: Time.zone.now)
-    project = notification.params[:project]
-    redirect_to user_project_path(project.author, project)
+    if notification.type == "NewAssignmentNotification"
+      assignment = notification.params[:assignment]
+      group = assignment.group
+      redirect_to group_assignment_path(group, assignment)
+    else
+      project = notification.params[:project]
+      redirect_to user_project_path(project.author, project)
```
Comment (tachyons) [inline app/controllers/users/noticed_notifications_controller.rb:20]: We need to clean this up and move logic to a service [object](https://www.toptal.com/ruby-on-rails/rails-service-objects-tutorial)

File: app/models/assignment.rb [hunk 17]
```diff
@@ -8,15 +8,30 @@ class Assignment < ApplicationRecord
...
+  has_noticed_notifications model_name: "NoticedNotification"
+
+  def notify_recipient
+    @assignment = Assignment.find(id)
+    group.group_members.each do |group_member|
+      NewAssignmentNotification.with(assignment: @assignment).deliver_later(group_member.user)
+    end
+  end
+
+  def cleanup_notification
```
Comment (tachyons) [inline app/models/assignment.rb:31]: Do we really need this? can we specify dependent option instead?

File: app/models/assignment.rb [hunk 18]
```diff
@@ -8,15 +8,30 @@ class Assignment < ApplicationRecord
...
   default_scope { order(deadline: :asc) }
   has_many :grades, dependent: :destroy
 
+  has_noticed_notifications model_name: "NoticedNotification"
+
+  def notify_recipient
+    @assignment = Assignment.find(id)
+    group.group_members.each do |group_member|
+      NewAssignmentNotification.with(assignment: @assignment).deliver_later(group_member.user)
+    end
```
Comment (tachyons) [inline app/models/assignment.rb:28]:
```suggestion
    group.group_members.each do |group_member|
      NewAssignmentNotification.with(assignment: self).deliver_later(group_member.user)
    end
```

Comment (tachyons) [reply app/models/assignment.rb:28]: prefer using self to avoid extra DB query

**Review by tachyons — CHANGES_REQUESTED** [review]

File: app/assets/stylesheets/users.scss [hunk 19]
```diff
@@ -468,7 +468,8 @@
   position: relative;
 
   .unseen-active {
-    top: 22px;
+    right: -6px;
+    top: 21px;
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/assets/stylesheets/users.scss:472]: @VaibhavUpreti make a separate issue for this and the respective PR!

[the remark this answers is not in the document]
Comment (VaibhavUpreti) [reply app/assets/stylesheets/users.scss:472]: yeah sure!

File: app/controllers/users/noticed_notifications_controller.rb (hunk 16 above)
Comment (VaibhavUpreti) [reply app/controllers/users/noticed_notifications_controller.rb:20]: @tachyons I didn't quite get why do we need to put this in a service class? Here we are dealing with params, routing redirects ,.. controller specific things. Just as mentioned in the article you have shared.

When Should I Not Use a Service Object?
This one’s easy. I have these rules:

Does your code handle routing, params or do other controller-y things?
If so, don’t use a service object—your code belongs in the controller.

Are you trying to share your code in different controllers?
In this case, don’t use a service object—use a concern.

Is your code like a model that doesn’t need persistence?
If so, don’t use a service object. Use a non-ActiveRecord model instead.

Is your code a specific business action? (e.g., “Take out the trash,” “Generate a PDF using this text,” or “Calculate the customs duty using these complicated rules”)
In this case, use a service object. That code probably doesn’t logically fit in either your controller or your model.

Comment (tachyons) [reply app/controllers/users/noticed_notifications_controller.rb:20]: Basically notifications have several types, we are currently handling only two. Handling all those in this way will make the controller too big. We need to tidy this up

Comment (vedant-jain03) [reply app/controllers/users/noticed_notifications_controller.rb:20]: @VaibhavUpreti This will help in clear code!

**Comment by VaibhavUpreti** [comment]
@tachyons @vedant-jain03 I haven't used the function "notification_type" as I think routing is part of ActionController and not present in custom classes. So the current code handles routing in the controller itself and "notify" function works for now.

Is there any way we can use "redirect_to" method in our service class ... is so then we can keep this function else remove this.

File: app/controllers/users/noticed_notifications_controller.rb [hunk 20]
```diff
@@ -11,8 +11,15 @@ def index
...
-    redirect_to user_project_path(project.author, project)
+    case NoticedNotificationType.new(params).notify
+    when true
+      assignment = notification.params[:assignment]
+      group = assignment.group
+      redirect_to group_assignment_path(group, assignment)
+    when false
+      project = notification.params[:project]
+      redirect_to user_project_path(project.author, project)
+    end
```
Comment (tachyons) [inline app/controllers/users/noticed_notifications_controller.rb:22]: Create a service which returns redirect path and success, you can have something like this. 
```
class NotifyUser
  Result = Struct.new(:success, :redirection_path)
  def initialize(notification_type, notifiable)
  end
  
  def call
    return Struct.new(,)
  end
end
```

**Comment by VaibhavUpreti** [comment]
@tachyons made the controller "skinny" ... moved all the logic to the service class. Kindly review.

File: app/controllers/users/noticed_notifications_controller.rb [hunk 21]
```diff
@@ -11,8 +11,10 @@ def index
   def mark_as_read
     notification = NoticedNotification.find(params[:notification_id])
     notification.update(read_at: Time.zone.now)
-    project = notification.params[:project]
-    redirect_to user_project_path(project.author, project)
+    answer = NotifyUser.new(params).call
+    return redirect_to group_assignment_path(answer.first_param, answer.second) if answer.type == "new_assignment"
```
Comment (vedant-jain03) [inline app/controllers/users/noticed_notifications_controller.rb:15]: 1. Why introduce a new `type` in service when we can already compare with schema `type`.
2. Stick to `if-else` block as we have some other notification coming up which may have a different path other than `project` or `group`. 
Else LGTM @tachyons WDYT?
